pkg/deploy/controller/deployment/controller.go

```diff
@@ -209,6 +209,17 @@ func (c *DeploymentController) Handle(deployment *kapi.ReplicationController) er
 	}
 
 	if deployutil.CanTransitionPhase(currentStatus, nextStatus) || deploymentScaled {
+		// if we are going to transition to failed or complete and scale is non-zero, we'll check one more
+		// time to see if we are a test deployment to guarantee that we maintain the test invariant.
+		if deployment.Spec.Replicas != 0 {
+			switch nextStatus {
+			case deployapi.DeploymentStatusComplete, deployapi.DeploymentStatusFailed:
+				if config, err := deployutil.DecodeDeploymentConfig(deployment, c.codec); err == nil && config.Spec.Test {
```

Comment:
This is getting annoying. Can't we just add a test annotation in the deployment when it's created? Then you wouldn't need to deserialize at all... and it's all over the place in this controller.